libp2p/multiaddress.nim

```diff
@@ -727,6 +727,10 @@ proc init*(mtype: typedesc[MultiAddress]): MultiAddress =
   ## Initialize empty MultiAddress.
   result.data = initVBuffer()
 
+proc tcpEndPoint*(address: IpAddress, port: Port): MultiAddress =
```

Comment:
@zah sorry but why we need this old `IpAddress` utility procedure?

Reply:
It's used in the beacon node to keep the code surrounding configuration / NAT setup / etc shorter. IPv4/TCP is still the predominant internet protocol and it doesn't hurt to add some shortcuts for it.

BTW, I imagine that it's possible to create a nice type-safe API for building multi addresses that can look like this:

```nim
var
   someIp = parseIpAddress(...)
   somePort = Port(x)
   a1 = multiaddr(someIp / tcp / somePort)
   a2 = multiaddr(https / "www.github.com")
```

It's easy to make it disallow certain invalid combinations such as `unix / somePort`

Reply:
@zah `var address = initTAddress("127.0.0.1:5000")` and what the problem with `MultiAddress.init("/ip4/127.0.0.1/tcp/50000")` or its so different to remember?

Reply:
https://github.com/status-im/nim-libp2p/blob/master/libp2p/wire.nim this API also helps you to establish connections using MultiAddress.

Reply:
LibP2P is library which tries to abstract over transport and you trying to introduce API call which breaks the core idea of LibP2P.

Reply:
Also if you wish such initialization procedure, why it called so weird? why it not called as:
```nim
proc init*(t: typedesc[MultiAddress], address: IpAddress, port: Port): MultiAddress =
  MultiAddress.init("/ip4/" & $address & "/tcp/" & $port)
```

Reply:
I considered this signature for `init`, but it would have been wrong, because it also needs a Protocol parameter. I could have set the default to `IPPROTO_TCP`, but this would have also been a bit arbitrary and surprising. Using a "named constructor" is not that bad sometimes.

Otherwise, I don't think this is breaking any core idea of LibP2P. It just adds a new constructor for the `MultiAddress` type. The more constructors we have, the easier the code is to use. There should be another one taking `TransportAddress` for sure. It just happens so that the existing code in the beacon chain repo was using the address types from Nim's standard library and there is nothing wrong with that. The addition is quite natural, because I didn't even have to introduce a new import here.

Reply:
If you want to use stdlib's `IpAddress` please do not make exceptions and create `init` constructor, i dont see any reasons for `tcpEndPoint` here.

Reply:
Also `nim-beacon-chain` will work not using TCP, but its going to work using LibP2P, so i think its better to start using MultiAddress there too.

Reply:
`MultiAddress` values are already used there as evident to the `newDaemonApi` changes, but there are also usability considerations - we won't be asking the users to enter multi-address strings on the command-line.

Reply:
I'm not sure about users, but bootnodes addresses will have MultiAddress addresses just because program must be sure that public key can be verified. For example current Ethereum1 bootnode addresses looks like this:
```
enode://a979fb575495b8d6db44f750317d0f4622bf4c2aa3365d6af7c284339968eef29b69ad0dce72a4d8db5ebb4968de0e3bec910127f134779fbcb0cb6d3331163c@52.16.188.185:30303
enode://aa36fdf33dd030378a0168efe6ed7d5cc587fafa3cdd375854fe735a2e11ea3650ba29644e2db48368c46e1f60e716300ba49396cd63778bf8a818c09bded46f@13.93.211.84:30303
```

MultiAddress addresses will look like this:

```
/ipfs/a979fb575495b8d6db44f750317d0f4622bf4c2aa3365d6af7c284339968eef29b69ad0dce72a4d8db5ebb4968de0e3bec910127f134779fbcb0cb6d3331163c
/ipfs/aa36fdf33dd030378a0168efe6ed7d5cc587fafa3cdd375854fe735a2e11ea3650ba29644e2db48368c46e1f60e716300ba49396cd63778bf8a818c09bded46f
```
or
```
/ip4/13.93.211.84/tcp/30303/p2p/aa36fdf33dd030378a0168efe6ed7d5cc587fafa3cdd375854fe735a2e11ea3650ba29644e2db48368c46e1f60e716300ba49396cd63778bf8a818c09bded46f
```

Reply:
As you can see `IpAddress` is not helping here.

Reply:
Also other place where `IpAddress` is used is NAT address which must be not an IPv4/IPv6 address, but pair of `address:port` and this form of address are better working with `TransportAddress`.

Reply:
on this note, what happens if a MultiAddress is initialized with an unrecognised protocol? I hope it doesn't raise?

Reply:
It will raise
https://github.com/status-im/nim-libp2p/blob/master/libp2p/multiaddress.nim#L686-L688

Reply:
chatted with @cheatfate a little about this - the thought behind the question was that it would be nice to support forwards compatibility as far as possible, so that when new schemes come along the code can keep working will keep working without needing upgrades and that parsing or interpretation happens only when it's absolutely necessary. we identified that it might not be possible always to convert unknown text formats into binary, but it might be possible to forward unknown binary addresses on wire, for example